### PR TITLE
REPLAY- 1893: Implement bitmap downscaling

### PR DESCRIPTION
**Pull request by jonathanmos**
### What does this PR do?
Implement bitmap downscaling similar to what was done on iOS, aiming to end up with a bitmap of 10kb or less. We calculate the expected dimensions sizes for a perfect square with the required filesize, and then work out the dimensions of the resized bitmap from that.

Things to consider: 

- We are later compressing the bitmaps to webp and this gives additional savings in file size, but it's difficult to predict in advance how much the savings will be. Should we be taking this into account in the scaling of the raw bitmap?

Some numbers from testing:
Initial bitmap size: 26544 
Size of bitmap after downscaling: 9400
After webp compression: 3136

### Motivation
This pr will allow us to more accurately size the lrucache

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt [hunk 1]
```diff
@@ -31,4 +32,46 @@ internal class DrawableUtils(
         drawable.draw(canvas)
         return bitmap
     }
+
+    internal fun downscaleBitmapToApproxSize(
```
Comment (jonathanmos) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:36]: perhaps this should be moved to a new BitmapUtils class, because it doesn't actually involve drawables

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt [hunk 2]
```diff
@@ -62,8 +62,8 @@ class Base64Serializer private constructor(
     }
 
     @VisibleForTesting
-    internal fun isOverSizeLimit(bitmapSize: Int): Boolean =
-        bitmapSize > BITMAP_SIZE_LIMIT_BYTES
```
Comment (0xnm) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt:66]: probably this change can be reverted, we can even name argument here just `size`.

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt [hunk 3]
```diff
@@ -87,21 +87,24 @@ class Base64Serializer private constructor(
...
     @WorkerThread
-    private fun convertBmpToBase64(bitmap: Bitmap): String {
-        val byteArrayOutputStream = webPImageCompression.compressBitmapToStream(bitmap)
+    private fun convertBmpToBase64(initialBitmap: Bitmap): String {
+        var base64Result = ""
 
-        if (isOverSizeLimit(byteArrayOutputStream.size())) {
-            return ""
-        }
+        val scaledBitmap = drawableUtils.downscaleBitmapToApproxSize(initialBitmap)
```
Comment (0xnm) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt:93]: I would add what `approximate` means in this context, maybe it is worth adding the doc / explanation to the method?

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt [hunk 4]
```diff
@@ -87,21 +87,24 @@ class Base64Serializer private constructor(
     }
 
     @WorkerThread
-    private fun convertBmpToBase64(bitmap: Bitmap): String {
-        val byteArrayOutputStream = webPImageCompression.compressBitmapToStream(bitmap)
+    private fun convertBmpToBase64(initialBitmap: Bitmap): String {
```
Comment (0xnm) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt:90]: in this flow we create 2 bitmaps: first bitmap to draw content of drawable, second bitmap to downscale. I'm wondering if we can do this in 1 pass: can we create a canvas which will apply affine transformation (scale)?

In theory downscaling is just applying the matrix, so maybe we can create a downscaled bitmap straight away (we know size of the drawable, so we should be able to do it), and then provide a canvas which will apply the necessary transformation? This would greatly reduce memory usage.

**Review by 0xnm — COMMENTED** [review]
I left one comment wondering if we can use only 1 bitmap during the image conversion flow, I guess it should be possible.

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt [hunk 5]
```diff
@@ -87,21 +87,24 @@ class Base64Serializer private constructor(
...
-        try {
-            base64 = base64Utils.serializeToBase64String(byteArrayOutputStream)
-        } finally {
-            bitmap.recycle()
+        scaledBitmap?.let {
+            val byteArrayOutputStream = webPImageCompression.compressBitmapToStream(it)
+
+            try {
+                if (!isOverSizeLimit(byteArrayOutputStream.size())) {
+                    base64Result = base64Utils.serializeToBase64String(byteArrayOutputStream)
```
Comment (mariusc83) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt:100]: maybe we want a log here in case is over the limit ?

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt (hunk 1 above)
Comment (jonathanmos) [reply features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt]: In the end not necessary, since refactoring to perform the downscaling when the bitmap is first created removes this method

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt (hunk 5 above)
Comment (jonathanmos) [reply features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt]: I'm removing the method, since we now know the bitmap is already downscaled under the size limit, and compressing to webp can only make it smaller. We could also move the size limit method to later in order to test the serialized string for size, however, base64 adds a constant 30% to the bitmap size so we can already predict it and handle it by making the scaled bitmap size at least 30% smaller than the limit.

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/wrappers/BitmapWrapper.kt [hunk 6]
```diff
@@ -27,4 +27,25 @@ internal class BitmapWrapper {
             null
         }
     }
+
+    internal fun createScaledBitmap(
```
Comment (mariusc83) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/wrappers/BitmapWrapper.kt:31]: I wonder why do you need the `createBitmap` above ? As that one only delegates to this one.

File: features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtilsTest.kt [hunk 7]
```diff
@@ -105,26 +109,57 @@ internal class DrawableUtilsTest {
...
+            assertThat(it).isEqualTo(50)
+        }
+
+        assertThat(displayMetricsCaptor.firstValue).isEqualTo(mockDisplayMetrics)
+
+        assertThat(result).isEqualTo(mockBitmap)
+    }
+
+    @Test
+    fun `M return null W createBitmapFromDrawableOfApproxSize() { failed to create bmp }`() {
```
Comment (mariusc83) [inline features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtilsTest.kt:150]: I think you are missing the test where the `canvasWrapper` returns null ?

**Review by mariusc83 — APPROVED** [review]
LGTM

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt [hunk 8]
```diff
@@ -13,22 +13,67 @@ import android.util.DisplayMetrics
...
+            displayMetrics,
+            dimensions.first,
+            dimensions.second,
+            Bitmap.Config.ARGB_8888
+        )
             ?: return null
 
         val canvas = canvasWrapper.createCanvas(bitmap) ?: return null
         drawable.setBounds(0, 0, canvas.width, canvas.height)
         drawable.draw(canvas)
```
Comment (0xnm) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:46]: so how does writing the scaled bitmap works? I'm asking, because it seems we don't specify any transformation matrix. Say we have `BitmapDrawable` and its bitmap is 1000x1000, will it be properly recorded to the bitmap of size 100x100?

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt [hunk 9]
```diff
@@ -13,22 +13,67 @@ import android.util.DisplayMetrics
...
+                width = (maxSize * bitmapRatio).toInt()
+            }
+        }
+
+        return Pair(width, height)
+    }
+
+    internal companion object {
+        internal const val DESIRED_BITMAP_SIZE_IN_BYTES = 10240 // 10kb
+        private const val ARGB_8888_DEPTH = 4
```
Comment (0xnm) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:77]: a bit more descriptive naming:

```suggestion
        private const val ARGB_8888_PIXEL_SIZE_BYTES = 4
```

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt [hunk 10]
```diff
@@ -13,22 +13,67 @@ import android.util.DisplayMetrics
...
+        drawable: Drawable,
+        requestedSizeInBytes: Int
+    ): Pair<Int, Int> {
+        var width = if (drawable.intrinsicWidth <= 0) 1 else drawable.intrinsicWidth
+        var height = if (drawable.intrinsicHeight <= 0) 1 else drawable.intrinsicHeight
+        val sizeAfterCreation = width * height * ARGB_8888_DEPTH
+
+        if (sizeAfterCreation > requestedSizeInBytes) {
+            val bitmapRatio = width.toDouble() / height.toDouble()
+            val perfectSquareSize = (requestedSizeInBytes / ARGB_8888_DEPTH).toDouble()
```
Comment (0xnm) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:60]:
```suggestion
            val totalMaxPixels = (requestedSizeInBytes / ARGB_8888_DEPTH).toDouble()
```

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt [hunk 11]
```diff
@@ -13,22 +13,67 @@ import android.util.DisplayMetrics
...
         drawable.setBounds(0, 0, canvas.width, canvas.height)
         drawable.draw(canvas)
         return bitmap
     }
+
+    private fun getScaledWidthAndHeight(
+        drawable: Drawable,
+        requestedSizeInBytes: Int
+    ): Pair<Int, Int> {
+        var width = if (drawable.intrinsicWidth <= 0) 1 else drawable.intrinsicWidth
```
Comment (0xnm) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:54]: in which case `intrinsicWidth` can be 0 or negative? if we get `width` and `height` equal to 1, we are going to send this 1 pixel to backend? seems a bit strange

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt [hunk 12]
```diff
@@ -13,22 +13,67 @@ import android.util.DisplayMetrics
...
+            } else {
+                width = (maxSize * bitmapRatio).toInt()
+            }
+        }
+
+        return Pair(width, height)
+    }
+
+    internal companion object {
+        internal const val DESIRED_BITMAP_SIZE_IN_BYTES = 10240 // 10kb
```
Comment (0xnm) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:76]: maybe max bitmap size? because `desired` has a connotation that we may upscale bitmap if size is below

```suggestion
        internal const val MAX_BITMAP_SIZE_IN_BYTES = 10240 // 10kb
```

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt (hunk 8 above)
Comment (jonathanmos) [reply features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:46]: It seems to me that there's is automatic scaling applied. `Drawable.draw` is calling `drawBitmap`, and the comment for this method says: "Draw the specified bitmap, scaling/translating automatically to fill the destination rectangle."

Comment (0xnm) [reply features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:46]: I'm wondering if scaling also applies to other draw operation (`drawRect`, `drawCircle`, etc.) which take coordinates as parameters.

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt (hunk 11 above)
Comment (0xnm) [reply features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:54]: yes, I would pull it up probably, because if we don't have drawable size we cannot do anything anyway.

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/mapper/ImageButtonMapper.kt [hunk 13]
```diff
@@ -39,22 +43,58 @@ internal class ImageButtonMapper(
...
+            val imageWireframe = resolveImageWireframe(
+                view,
+                id,
+                bounds,
+                shapeStyle,
+                border,
+                drawable
+            )
+            wireframes.add(imageWireframe)
+        } else {
```
Comment (jonathanmos) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/mapper/ImageButtonMapper.kt:58]: This raises the question - is it better not to return a wireframe at all in this case, or should we return a shape with the bounds of the view?

Comment (mariusc83) [reply features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/mapper/ImageButtonMapper.kt:58]: we should not return as it will not have any visible effect in the player.

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt (hunk 8 above)
Comment (mariusc83) [reply features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:46]: This can be tested inside the sample app by adding these particular cases.

Comment (jonathanmos) [reply features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:46]: I've added some shapes to the sample app

File: features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/mapper/ImageButtonMapper.kt [hunk 14]
```diff
@@ -18,15 +20,17 @@ internal class ImageButtonMapper(
...
     webPImageCompression = webPImageCompression,
     base64Serializer = base64Serializer,
     uniqueIdentifierGenerator = uniqueIdentifierGenerator
 ) {
     override fun map(
         view: ImageButton,
         mappingContext: MappingContext
-    ): List<MobileSegment.Wireframe.ImageWireframe> {
+    ): List<MobileSegment.Wireframe> {
+        val wireframes = mutableListOf<MobileSegment.Wireframe>()
```
Comment (0xnm) [inline features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/mapper/ImageButtonMapper.kt:32]: minor: it is better to move declaration closer to the usage site (L46)

**Review by 0xnm — APPROVED** [review]
